scripts/asr_language_modeling/ngram_lm/eval_beamsearch_ngram_transducer.py

```diff
@@ -13,257 +13,420 @@
+           beam_width=[<list of the beam widths, separated with commas>] \
+           beam_alpha=[<list of the beam alphas, separated with commas>] \
+           preds_output_folder=<optional folder to store the predictions> \
+           probs_cache_file=null \
+           decoding_strategy=<greedy_batch or maes decoding>
+           maes_prefix_alpha=[<list of the maes prefix alphas, separated with commas>] \
+           maes_expansion_gamma=[<list of the maes expansion gammas, separated with commas>] \
+           hat_subtract_ilm=<in case of HAT model: subtract internal LM or not> \
+           hat_ilm_weight=[<in case of HAT model: list of the HAT internal LM weights, separated with commas>] \
+           ...
+
+
+# Grid Search for Hyper parameters
+
+For grid search, you can provide a list of arguments as follows -
+
+           beam_width=[4,8,16,....] \
+           beam_alpha=[-2.0,-1.0,...,1.0,2.0] \
+
 # You may find more info on how to use this script at:
 # https://docs.nvidia.com/deeplearning/nemo/user-guide/docs/en/main/asr/asr_language_modeling.html
 
+"""
+
 
-import argparse
 import contextlib
 import json
 import os
 import pickle
+from dataclasses import dataclass, field, is_dataclass
+from pathlib import Path
+from typing import List, Optional
 
 import editdistance
+import numpy as np
 import torch
-from omegaconf import OmegaConf
+from omegaconf import MISSING, OmegaConf
+from sklearn.model_selection import ParameterGrid
 from tqdm.auto import tqdm
 
 import nemo.collections.asr as nemo_asr
-from nemo.collections.asr.metrics.rnnt_wer_bpe import RNNTBPEDecodingConfig
+from nemo.collections.asr.parts.submodules import rnnt_beam_decoding
+from nemo.core.config import hydra_runner
 from nemo.utils import logging
 
+# fmt: off
+
+
+@dataclass
+class EvalBeamSearchNGramConfig:
+    """
+    Evaluate an ASR model with beam search decoding and n-gram KenLM language model.
+    """
+    # # The path of the '.nemo' file of the ASR model or the name of a pretrained model (ngc / huggingface)
+    nemo_model_file: str = MISSING
+
+    # File paths
+    input_manifest: str = MISSING  # The manifest file of the evaluation set
+    kenlm_model_file: Optional[str] = None  # The path of the KenLM binary model file
+    preds_output_folder: Optional[str] = None  # The optional folder where the predictions are stored
+    probs_cache_file: Optional[str] = None  # The cache file for storing the logprobs of the model
+
+    # Parameters for inference
+    acoustic_batch_size: int = 128  # The batch size to calculate log probabilities
+    beam_batch_size: int = 128  # The batch size to be used for beam search decoding
+    device: str = "cuda"  # The device to load the model onto to calculate log probabilities
+    use_amp: bool = False  # Whether to use AMP if available to calculate log probabilities
+
+    # The decoding scheme to be used for evaluation
+    decoding_strategy: str = "greedy_batch" # ["greedy_batch", "maes"]
```

Comment:
Revert these args below specific to decoding strategy (why is this greedy_batch for a beam search script btw) and use the values from BeamRNNTInferConfig here - similar to how it is done in the CTC beam search example. Use that config as you need. Dont redo maes, ILM etc arguments.

The only exception is alpha, beta and beam_width - to stay consistent with CTC beam search framework.